core/rawdb/freezer_table.go

```diff
@@ -364,6 +371,122 @@ func (t *freezerTable) repair() error {
 	return nil
 }
 
+// repairIndex validates the integrity of the index file. According to the design,
+// the initial entry in the file denotes the earliest data file along with the
+// count of deleted items. Following this, all subsequent entries in the file must
+// be in order. This function identifies any corrupted entries and truncates items
+// occurring after the corruption point.
+//
+// The corruption can be made because of the power failure. As in the Linux kernel,
+// the file metadata update and data update are not necessarily performed at the
+// same time. Typically, the metadata will be flushed/journaled ahead of the file
+// data. Therefore, we make the pessimistic assumption that the file is first
+// extended with invalid "garbage" data (normally zero bytes) and that afterwards
+// the correct data replaces the garbage. As all the items in index file are
+// supposed to be in-order, the leftover garbage must be truncated before the
+// index data is utilized.
+//
+// It's important to note an exception that's unfortunately undetectable: when
+// all index entries in the file are zero. Distinguishing whether they represent
+// leftover garbage or if all items in the table have zero size is impossible.
+// In such instances, the file will remain unchanged to prevent potential data
+// loss or misinterpretation.
+func (t *freezerTable) repairIndex() error {
+	// Retrieve the file sizes and prepare for validation
+	stat, err := t.index.Stat()
+	if err != nil {
+		return err
+	}
+	size := stat.Size()
+
+	var (
+		start      = time.Now()
+		batchSize  = int64(indexEntrySize * 1024 * 1024)
+		buffer     = make([]byte, batchSize) // pre-allocate for batch reading
+		prev       indexEntry
+		head       indexEntry
+		readOffset int64
+		consumed   int64
+
+		read = func(offset int64) (indexEntry, error) {
+			if offset+indexEntrySize > size {
+				return indexEntry{}, fmt.Errorf("slice bounds out of range, offset: %d, size: %d", offset, size)
+			}
+			if offset >= readOffset {
+				n, err := t.index.ReadAt(buffer, readOffset)
+				if err != nil && !errors.Is(err, io.EOF) {
+					return indexEntry{}, err
+				}
+				expect := batchSize
+				if size-readOffset < batchSize {
+					expect = size - readOffset
+				}
+				if expect != int64(n) {
+					return indexEntry{}, fmt.Errorf("failed to read from index, want: %d, got: %d", expect, n)
+				}
```

Comment:
This whole chunk of code, I don't really get it. For buffering, couldn't you just use `bufio.NewReader` and let that take care of some of the complexities here? 
Or does this achieve something more than just buffered reads?

Reply:
yeah, originally i thought it's a bit hacky as `bufio.Reader` will advance the file cursor whenever read is performed.
But i realized that we always use `indexfile.ReadAt` in other places, so it should be fine.